XenAdmin/Dialogs/VtmpManagementDialog.cs

```diff
@@ -0,0 +1,333 @@
+                using (var dlg = new ErrorDialog(msg))
+                    dlg.ShowDialog(Parent);
+
+                return;
+            }
+
+            var action = new NewVtpmAction(_vm.Connection, _vm);
+            using (var dlg = new ActionProgressDialog(action, ProgressBarStyle.Marquee))
+                dlg.ShowDialog(this);
+        }
+
+        private void RemoveVtpm(VTPM vtpm)
+        {
+            HideTooltip();
+
+            using (var dlg = new WarningDialog(Messages.VTPM_REMOVE_WARNING, ThreeButtonDialog.ButtonYes, ThreeButtonDialog.ButtonNo))
+            {
+                if (dlg.ShowDialog() != DialogResult.Yes)
+                    return;
+            }
+
+            if (!_vm.Connection.Session.IsLocalSuperuser && !Registry.DontSudo &&
+                _vm.Connection.Session.Roles.All(r => r.name_label != Role.MR_ROLE_VM_ADMIN))
+            {
+                var currentRoles = _vm.Connection.Session.Roles;
+                currentRoles.Sort();
+
+                var msg = string.Format(Messages.VTPM_RBAC_RESTRICTION_REMOVE, currentRoles[0].FriendlyName(),
+                    Role.FriendlyName(Role.MR_ROLE_VM_ADMIN));
+
+                using (var dlg = new ErrorDialog(msg))
+                    dlg.ShowDialog(Parent);
+
+                return;
+            }
+
+            var action = new RemoveVtpmAction(vtpm.Connection, vtpm);
+            using (var dlg = new ActionProgressDialog(action, ProgressBarStyle.Marquee))
+                dlg.ShowDialog(this);
+        }
+
+        private void ResizeVerticalTabs()
+        {
+            int maxHeight = splitContainer.Panel1.Height - toolTipContainerAdd.Height;
+            verticalTabs.Height = Math.Min(maxHeight, verticalTabs.Items.Count * verticalTabs.ItemHeight);
+            toolTipContainerAdd.Top = verticalTabs.Top + verticalTabs.Height;
+        }
+
+        private void ShowTooltip(Point location)
+        {
+            if (!_toolTipVisible)
+            {
+                Cursor = Cursors.Hand;
+                var msg = _vm.CanRemoveVtpm(out var cannotReason) ? Messages.VTPM_REMOVE : cannotReason;
+
+                if (!string.IsNullOrEmpty(cannotReason))
+                {
+                    toolTip.Show(msg, verticalTabs, location.X, location.Y + 20);
+                    _toolTipVisible = true;
+                }
+            }
+        }
+
+        private void HideTooltip()
+        {
+            toolTip.Hide(verticalTabs);
+            _toolTipVisible = false;
+            Cursor = Cursors.Default;
+        }
+
+        #region Event handlers
+
+        private void _vm_PropertyChanged(object sender, PropertyChangedEventArgs e)
+        {
+            if (!(sender is VM))
+                return;
+
+            if (e.PropertyName != "power_state" && e.PropertyName != "VTPMs")
+                return;
+
+            Program.Invoke(this, RefreshTabs);
+        }
+
+
+        private void splitContainer_Panel1_Resize(object sender, EventArgs e)

```

Comment:
very nit, but there's a few methods here that are separated by two empty lines instead of one. If you're making other changes, it'd be nice to also normalise this.